project4/name_server/nameserver.cpp

```diff
@@ -1,41 +1,276 @@
+  // be the last in the ring. if the request does not have successor info but
+  // has predecessor info by the time it reaches back to bootstrap, we know the
+  // entering nameserver will be the first in the ring.
+}
 
 void Nameserver::HandleRequest(
-    const consistent_hash_msgs::ExitInformation& request) {}
+    const consistent_hash_msgs::ExitInformation& request) {
+  // take over the exiting server's key-vals
+  bounds_.second = request.upper_bounds();
+  for (int i = 0; i < request.keys_size(); i++) {
+    std::pair<int, std::string> pair(request.keys(i), request.values(i));
+    pairs_.insert(pair);
+  }
+
+  // update predecessor info
+  predecessor_.port = request.predecessor_info().port();
+  predecessor_.hostname = request.predecessor_info().ip();
+  predecessor_id_ = request.predecessor_info().id();
+}
 
 void Nameserver::HandleRequest(
-    const consistent_hash_msgs::UpdatePredecessorRequest& request) {}
+    consistent_hash_msgs::UpdatePredecessorRequest& request,
+    message_passing::Endpoint source) {
+  // update predecessor info
+  predecessor_.port = request.predecessor_info().port();
+  predecessor_.hostname = source.hostname;
+  predecessor_id_ = request.predecessor_info().id();
+
+  // give lower half of it's key-val range to the entering nameserver
+  consistent_hash_msgs::UpdatePredecessorResponse res;
+  res.set_lower_bounds(bounds_.first);
+  res.set_upper_bounds(bounds_.second / 2);
+  uint count = 0;
+  for (std::pair<uint, std::string> pair : pairs_) {
+    res.set_keys(count, pair.first);
+    res.set_values(count, pair.second);
+    pairs_.erase(pair.first);
+    if (count == (bounds_.second / 2) - bounds_.first) {
+      break;
+    }
+  }
+  bounds_.first = bounds_.second / 2 + 1;
+
+  client_ =
+      std::make_unique<message_passing::Client>(threadpool_, predecessor_);
+  if (client_->Send(res) < 0) {
+    // error
+  }
+}
 
 void Nameserver::HandleRequest(
-    const consistent_hash_msgs::UpdateSuccessorRequest& request) {}
+    const consistent_hash_msgs::UpdateSuccessorRequest& request) {
+  // update successor info
+  successor_.hostname = request.successor_info().ip();
+  successor_.port = request.successor_info().port();
+  successor_id_ = request.successor_info().id();
+}
 
 void Nameserver::HandleRequest(
     const consistent_hash_msgs::UpdatePredecessorResponse& request) {}
 
-void Nameserver::HandleRequest(
-    const consistent_hash_msgs::LookUpResult& request) {}
+void Nameserver::HandleRequest(consistent_hash_msgs::LookUpResult& request) {
+  if (request.key() <= bounds_.second && request.key() >= bounds_.first) {
+    // key-val resides in this nameserver
+    auto itr = pairs_.find(request.key());
+    if (itr != pairs_.end()) {
+      // value found
+      request.set_value(itr->second);
+      request.set_id(id_);
+    }
+  }
+  // denote this server as contacted
+  auto ids = request.server_ids();
+  request.set_server_ids(ids.size(), id_);
```

Comment:
I think the method that you want here is `add_server_ids()`. `set_server_ids()` might actually be undefined behavior since your index is technically out-of-bounds.